[ADT] Move llvm::identity into IndexedMap (NFC)

kazutakahirata opened the pull request:
This patch moves llvm::identity to IndexedMap for two reasons:

- llvm::identity is used only IndexedMap.
- llvm::identity is not suitable for general use as it is not quite
  the same as std::identity despite the comments in identity.h.

Also, this patch renames the class to IdentityIndex and places it in
the "detail" namespace.

kazutakahirata committed:
[ADT] Modernize IndexedMap to use llvm::identity_cxx20 (NFC)

The legacy llvm::identity differs from std::identity in C++20.
llvm::identity is a template struct with an ::argument_type member. In
contrast, llvm::identity_cxx20 (and std::identity) is a non-template
struct with a templated call operator and no ::argument_type.

This patch modernizes llvm::IndexedMap by updating its default functor
to llvm::identity_cxx20.  A new template parameter IndexT now
fulfills the role of ::argument_type.

To avoid modifying numerous existing IndexedMap uses with custom
functors (e.g., VirtReg2IndexFunctor), the detail::DeduceIndexType
helper automatically deduces IndexT.  It deduces ::argument_type for
legacy functors and unsigned for llvm::identity_cxx20.

kuhar reviewed (commented):
Shouldn't we retain `llvm::identity` for cases like this? It seems much cleaner to me.

kazutakahirata committed:
[ADT] Modernize IndexedMap with llvm::identity_cxx20 (NFC)

C++ is evolving in the direction of removing type alias
"argument_type" from functors.

This patch changes the default functor type to llvm::identity_cxx20,
backported from C++20, while templatizing those methods that use
toIndex_ with IndexT.  This way we no longer need to take
argument_type from ToIndexT.

Now, identity_cxx20()(int) is of type int, causing mixed sign
comparisons, so this patch first assigns the result to a temporary
variable of type unsigned before feeding it to assert.

kazutakahirata commented:
> Shouldn't we retain `llvm::identity` for cases like this? It seems much cleaner to me.

@kuhar Well, C++ is moving away from `argument_type`.  How about removing the dependence on `argument_type` completely?  Please take a look at the latest iteration.  Thanks!

kazutakahirata committed:
[ADT] Move llvm::identity into IndexedMap (NFC)

This patch moves llvm::identity to IndexedMap for two reasons:

- llvm::identity is used only IndexedMap.
- llvm::identity is not suitable for general use as it is not quite
  the same as std::identity despite the comments in identity.h.

Also, this patch renames the class to IdentityIndex and places it in
the "detail" namespace.

kazutakahirata committed:
Merge branch 'main' into cleanup_20251021_ADT_IndexedMap_identity

kazutakahirata committed:
Address a comment.
